actions/publish-github-pages/entrypoint.py

```diff
@@ -0,0 +1,237 @@
+    STAGING_DIR = STAGING_DIR.rstrip("/")
+
+def checkout_repos():
+
+    # Check out:
+    # 1) Essential Hugo build tools
+    Repo.clone_from(HUGO_REPO, GITHUB_PAGES_HUGO_PATH)
+
+    # 2) Site-specific documentation
+    Repo.clone_from(GITHUB_PAGES_REPO, GITHUB_PAGES_PULL_PATH, branch=GITHUB_PAGES_PULL_BRANCH)
+
+    # 3) The GitHub Pages deployment branch for this site
+    Repo.clone_from(GITHUB_PAGES_REPO, GITHUB_PAGES_PUSH_PATH, branch=GITHUB_PAGES_PUSH_BRANCH)
+
+def rearrange_files(content_base_path):
+
+    # rename all `README.md` to `_index.md` (hugo convention)
+    for root, dirs, files in os.walk(content_base_path):
+        for local_file in files:
+            if local_file=="README.md":
+                os.rename( os.path.join(root, local_file), os.path.join(root, "_index.md") )
+
+    # folders with no subfolders, and only a single md file (usually `_index.md`): `mv foobar/_index.md foobar.md`
+    hierarchy_modified = True # init value
+    while hierarchy_modified:
+        hierarchy_modified = False
+        for root, dirs, files in os.walk(content_base_path):
+            if root != content_base_path:
+                if not len(dirs):
+                    markdown_files = [potential_md_file for potential_md_file in files if ".md" in potential_md_file]
+                    if len(markdown_files) == 1:
+                        os.rename( os.path.join(root, markdown_files[0]), \
+                                   os.path.join(root.rsplit('/',1)[0], root.rsplit('/',1)[1] + ".md") )
+                        rmtree(root)
+                        hierarchy_modified = True
+                    elif len(markdown_files) == 0:
+                        rmtree(root)
+                        hierarchy_modified = True
+
+def insert_frontmatter(file_path, weight=1):
+
+    # check for frontmatter in this file
+    frontmatter_flag = True
+    with open(file_path, "r") as markdown_file:
+        line = markdown_file.readline()
+        if not re.match("^---", line):
+            frontmatter_flag = False
+
+    # if the frontmatter wasn't found, create it by finding the markdown header
+    # and using it as the core of the frontmatter
+    if not frontmatter_flag:
+        title = ""
+        with open(file_path, "r") as markdown_file:
+            for line in markdown_file:
+                match_results = re.match("^\s*#+\s*(.*)$", line)
+                if match_results:
+                    title = match_results.group(1)
+                    break
+
+        # write out file with frontmatter
+        with open(file_path, "r") as markdown_file:
+            input_file = markdown_file.read()
+        with open(file_path, "w") as markdown_file:
+            if not title:
+                title="default_title"
+            front_matter_string = "---\n" + \
+                                  f"title: \"{title}\"\n" + \
+                                  "description: \"No description available.\"\n" + \
```

Comment:
Since the description will be present on every page, let's not put this here. 

![image](https://user-images.githubusercontent.com/52489/112791435-6dc15880-9016-11eb-839a-fb2c39bf599e.png)

It will look ugly if we always say "No description available"